proposals/simd-0182-conditional-cu-metering.md

```diff
@@ -0,0 +1,114 @@
+
+However, when an exception is thrown during the execution of a Basic Block
+(e.g., a null memory dereference or other faults), determining the exact number
+of CUs consumed up to the point of failure requires additional effort. For
+instance, the Agave client implements a mechanism that tracks the Instruction
+Pointer (IP) or Program Counter (PC) to backtrack and estimate the CUs consumed
+when an exception occurs. More details on this mechanism can be found
+[here](https://github.com/solana-labs/rbpf/blob/57139e9e1fca4f01155f7d99bc55cdcc25b0bc04/src/jit.rs#L267).
+
+While this approach is effective, it introduces additional work and complexity.
+These mechanisms are often implementation-specific, and requiring all clients to
+track the exact number of executed sBPF instructions for consensus is costly and
+unnecessary. Such precision is not essential for protocol-level consensus,
+especially since these cases are rare.
+
+Instead of mandating implementation-specific work to handle exceptions, we
+propose the following clarification in the protocol:
+
+- For successful execution of a Basic Block (i.e., the block exits at the last
+ sBPF instruction), the deterministic CU cost of the block will be charged to
+the transaction’s CU meter. This ensures that CU consumption for successful
+transactions is accurately accounted for.
+- In the event of irregular failure, where execution aborts from the middle of
+basic block, the requested CUs for the transaction will be charged to the CU
+meter. This allows for a simple and efficient fallback mechanism that avoids the
+need for tracking the exact number of executed instructions up to the point of
+failure.
+
+By adopting this approach, the protocol avoids the overhead of requiring precise
+instruction-level CU tracking for transactions that fail. Instead, the requested
+CU limit of the transaction will be used, simplifying the handling of
+irregularly failed transactions while still maintaining consensus.
+
+## Alternatives Considered
+
+None
+
+## New Terminology
+
+- [Basic Block](https://en.wikipedia.org/wiki/Basic_block): In the context of
+  JIT execution and BPF processing, a Basic Block is a sequence of BPF
+instructions that forms a single, linear flow of control with no loops or
+conditional branches except for the entry and exit points. It represents a
+segment of code where execution starts at the first instruction and proceeds
+sequentially through to the last instruction without deviation. The Basic Block
+is characterized by its predictable execution path, allowing for efficient
+budget checks and optimizations, as its Compute Unit (CU) cost can be determined
+before execution and verified at the end of the block.
+
+- Irregular transaction failure: A rare case that a Transaction execution aborts
+in the middle of executing basic block, results in consuming all requested CUs.
+
+## Detailed Design
+
+If VM execution returns any error except `SyscallError`, transaction's CU meter
```

Comment:
I see that with direct mapping enabled, we convert `EbpfError::AccessViolation` into the `SyscallError` variant here: https://github.com/anza-xyz/agave/blob/af0ed22174999cb62579a0621f6b274c85ebf267/programs/bpf_loader/src/lib.rs#L1501

I am assuming that this access violation should also deplete compute units and should not be considered as an exception right?

Reply:
@Lichtso @alessandrod

Reply:
Yes, same reasoning as: https://github.com/solana-foundation/solana-improvement-documents/pull/182#discussion_r1827050882

Reply:
Should consume all cu's or not?

Reply:
Yes, we should. This is not caused by a syscall so the exact instruction counter would have to be calculated otherwise.

Reply:
Just want to call out that this type of access violation is probably going to be fairly common when direct mapping is enabled. So it's not exactly true to say that these types of errors are "rare, exceptional situations" or "A rare case" as we say in the SIMD. We should update the SIMD to say that these types of failures are "less common" rather than rare at least.

Reply:
[e8c74ac](https://github.com/solana-foundation/solana-improvement-documents/pull/182/commits/e8c74ac084a51587f405b07ea90df90b716f01f8)

Reply:
> I see that with direct mapping enabled, we convert EbpfError::AccessViolation into the SyscallError variant here: https://github.com/anza-xyz/agave/blob/af0ed22174999cb62579a0621f6b274c85ebf267/programs/bpf_loader/src/lib.rs#L1501

> I am assuming that this access violation should also deplete compute units and should not be considered as an exception right?

@tao-stones this should probably be made explicit in the SIMD. Maybe we can say that all EBPF errors _from the vm_ are treated as irregular.

Reply:
Mmm, imo, SIMD's text `If VM execution returns any error except SyscallError, ....` is clearer.

Reply:
I don't think that a spec should be based on "arbitrary" rust types.
We should clarify what's the expected behavior, for example if there's a memory issue inside a syscall, is that a Syscall or Epbf error?

I agree that if we enter a syscall and that fails, we know the exact number of CUs consumed, and so there's no need to deplete them all. I'm not sure if checking the SyscallError type is the correct implementation though. (In the example above, imo a memory issue inside a syscall should NOT deplete, however I don't know if the error thrown is  SyscallError or EbpfError)